// @metaplex-foundation/mpl-candy-guard/dist/src/generated/accounts/AllowListProof.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.allowListProofBeet = exports.AllowListProof = exports.allowListProofDiscriminator = void 0;
const beet = __importStar(require("@metaplex-foundation/beet"));
const web3 = __importStar(require("@solana/web3.js"));
const beetSolana = __importStar(require("@metaplex-foundation/beet-solana"));
exports.allowListProofDiscriminator = [19, 122, 207, 114, 207, 43, 233, 148];
class AllowListProof {
    constructor(timestamp) {
        this.timestamp = timestamp;
    }
    static fromArgs(args) {
        return new AllowListProof(args.timestamp);
    }
    static fromAccountInfo(accountInfo, offset = 0) {
        return AllowListProof.deserialize(accountInfo.data, offset);
    }
    static async fromAccountAddress(connection, address) {
        const accountInfo = await connection.getAccountInfo(address);
        if (accountInfo == null) {
            throw new Error(`Unable to find AllowListProof account at ${address}`);
        }
        return AllowListProof.fromAccountInfo(accountInfo, 0)[0];
    }
    static gpaBuilder(programId = new web3.PublicKey('Guard1JwRhJkVH6XZhzoYxeBVQe872VH6QggF4BWmS9g')) {
        return beetSolana.GpaBuilder.fromStruct(programId, exports.allowListProofBeet);
    }
    static deserialize(buf, offset = 0) {
        return exports.allowListProofBeet.deserialize(buf, offset);
    }
    serialize() {
        return exports.allowListProofBeet.serialize({
            accountDiscriminator: exports.allowListProofDiscriminator,
            ...this,
        });
    }
    static get byteSize() {
        return exports.allowListProofBeet.byteSize;
    }
    static async getMinimumBalanceForRentExemption(connection, commitment) {
        return connection.getMinimumBalanceForRentExemption(AllowListProof.byteSize, commitment);
    }
    static hasCorrectByteSize(buf, offset = 0) {
        return buf.byteLength - offset === AllowListProof.byteSize;
    }
    pretty() {
        return {
            timestamp: (() => {
                const x = this.timestamp;
                if (typeof x.toNumber === 'function') {
                    try {
                        return x.toNumber();
                    }
                    catch (_) {
                        return x;
                    }
                }
                return x;
            })(),
        };
    }
}
exports.AllowListProof = AllowListProof;
exports.allowListProofBeet = new beet.BeetStruct([
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['timestamp', beet.i64],
], AllowListProof.fromArgs, 'AllowListProof');
