// @metaplex-foundation/js/dist/esm/plugins/candyMachineModule/guards/tokenGate.mjs
import { tokenGateBeet } from '@metaplex-foundation/mpl-candy-guard';
import { mapSerializer, createSerializerFromBeet } from '../../../types/Serializer.mjs';
import { token } from '../../../types/Amount.mjs';

/**
 * The tokenGate guard restricts minting to token holders
 * of a specified mint account. The `amount` determines
 * how many tokens are required.
 *
 * This object defines the settings that should be
 * provided when creating and/or updating a Candy
 * Machine if you wish to enable this guard.
 *
 * @see {@link TokenGateGuardMintSettings} for more
 * information on the mint settings of this guard.
 */

/** @internal */
const tokenGateGuardManifest = {
  name: 'tokenGate',
  settingsBytes: 40,
  settingsSerializer: mapSerializer(createSerializerFromBeet(tokenGateBeet), settings => ({ ...settings,
    amount: token(settings.amount)
  }), settings => ({ ...settings,
    amount: settings.amount.basisPoints
  })),
  mintSettingsParser: ({
    metaplex,
    settings,
    mintSettings,
    payer,
    programs
  }) => {
    var _mintSettings$tokenAc;

    const tokenAccount = (_mintSettings$tokenAc = mintSettings === null || mintSettings === void 0 ? void 0 : mintSettings.tokenAccount) !== null && _mintSettings$tokenAc !== void 0 ? _mintSettings$tokenAc : metaplex.tokens().pdas().associatedTokenAccount({
      mint: settings.mint,
      owner: payer.publicKey,
      programs
    });
    return {
      arguments: Buffer.from([]),
      remainingAccounts: [{
        isSigner: false,
        address: tokenAccount,
        isWritable: false
      }]
    };
  }
};

export { tokenGateGuardManifest };
