// @metaplex-foundation/js/dist/cjs/plugins/programModule/ProgramClient.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.cjs');
var SdkError = require('../../errors/SdkError.cjs');

/**
 * @group Modules
 */
class ProgramClient {
  constructor(metaplex) {
    _rollupPluginBabelHelpers.defineProperty(this, "programs", []);

    this.metaplex = metaplex;
  }

  register(program) {
    this.programs.unshift(program);
  }

  all(overrides = []) {
    return [...overrides, ...this.programs];
  }

  allForCluster(cluster, overrides = []) {
    return this.all(overrides).filter(program => {
      var _program$clusterFilte, _program$clusterFilte2;

      return (_program$clusterFilte = (_program$clusterFilte2 = program.clusterFilter) === null || _program$clusterFilte2 === void 0 ? void 0 : _program$clusterFilte2.call(program, cluster)) !== null && _program$clusterFilte !== void 0 ? _program$clusterFilte : true;
    });
  }

  allForCurrentCluster(overrides = []) {
    return this.allForCluster(this.metaplex.cluster, overrides);
  }

  get(nameOrAddress, overrides = []) {
    const programs = this.allForCurrentCluster(overrides);
    const program = typeof nameOrAddress === 'string' ? programs.find(program => program.name === nameOrAddress) : programs.find(program => program.address.equals(nameOrAddress));

    if (!program) {
      throw new SdkError.ProgramNotRecognizedError(nameOrAddress, this.metaplex.cluster);
    }

    return program;
  }

  getGpaBuilder(nameOrAddress, overrides = []) {
    const program = this.get(nameOrAddress, overrides);

    if (!program.gpaResolver) {
      throw new SdkError.MissingGpaBuilderError(program);
    }

    return program.gpaResolver(this.metaplex);
  }

}

exports.ProgramClient = ProgramClient;
